osquery/events/darwin/endpointsecurity_fim.cpp

```diff
@@ -158,27 +158,33 @@ void EndpointSecurityFileEventPublisher::configure() {
           }
         });
 
-        // Invert muting for target paths, now any calls to mute path APIs will
-        // select instead of mute
-        es_invert_muting(es_file_client_, ES_MUTE_INVERSION_TYPE_TARGET_PATH);
-        // select only the paths we want, recommended best practice to call
-        // unmute on target paths, before calling "inverted" mute APIs
-        es_unmute_all_target_paths(es_file_client_);
-        for (auto p : file_paths_) {
-          if (std::find(exclude_paths_.begin(), exclude_paths_.end(), p) ==
-              exclude_paths_.end()) {
-            // p is not one of the excluded_paths, we monitor
-            auto result = isDirectory(p).ok()
-                              ? es_mute_path(es_file_client_,
-                                             p.c_str(),
-                                             ES_MUTE_PATH_TYPE_TARGET_PREFIX)
-                              : es_mute_path(es_file_client_,
-                                             p.c_str(),
-                                             ES_MUTE_PATH_TYPE_TARGET_LITERAL);
-            if (result == ES_RETURN_SUCCESS) {
-              VLOG(1) << "Monitoring path: " << p;
-            } else {
-              VLOG(1) << "Error while trying to monitor path: " << p;
+        // first check the mute inversion status for target paths
+        // only call the mute inversion APIs if we are not already "inverted"
+        if (es_muting_inverted(es_file_client_,
+                               ES_MUTE_INVERSION_TYPE_TARGET_PATH) ==
+            ES_MUTE_NOT_INVERTED) {
+          // Invert muting for target paths, now any calls to mute path APIs
+          // will select instead of mute
+          es_invert_muting(es_file_client_, ES_MUTE_INVERSION_TYPE_TARGET_PATH);
+          // select only the paths we want, recommended best practice to call
+          // unmute on target paths, before calling "inverted" mute APIs
+          es_unmute_all_target_paths(es_file_client_);
```

Comment:
I haven't yet tested but looking at the logic of this, the first time the configure function gets called, mute logic gets inverted **and** everything gets muted, then only paths in `file_paths` gets (un)muted.

On a second configure, if `file_paths` changes, nothing happens because the API inversion has already happened and the if skips the whole (un)muting logic.
So shouldn't the if only cover the call to `es_invert_muting`, while the rest is still being run again, so that it reacts to configure changes?

I assume that `es_unmute_all_target_paths` does what it says, and basically clears all the inclusions (when inverted), so that later `es_mute_path` can update such inclusions with the new `file_paths` state.

Reply:
And actually, even the prior logic connected to the flags, it seems that it will have a similar problem, where it will continue to accumulate executables muted, instead of resetting and updating to the new state as portrayed by the flags values.

Reply:
Updated, thanks @Smjert 

> the first time the configure function gets called, mute logic gets inverted and everything gets muted, then only paths in file_paths gets (un)muted.

> On a second configure, if file_paths changes, nothing happens because the API logic inversion has already happened and the if skips the whole (un)muting logic.

That is correct. I was being overly cautious, with the `if` block covering more.

The `if` now covers `es_invert_muting` and `es_unmute_all_target_paths`, which is what we want and now reacts to configure changes.
For more context, every ES client has a built in "default mute set" which contains system and platform binaries. When we "invert" muting, the default mute set (which has things like `bluetoothd` and `wifid` among others) gets selected, which is not what we want. As apple describes it: 

```
* @note Mute inversion does NOT clear the default mute set.
 * When a new `es_client_t` is created certain paths are muted by default.
 * This is known as "the default mute set".
 * The default mute set exists to protect ES clients from deadlocks, and to prevent watchdog timeout panics.
 * Creating a new client and calling `es_invert_muting(c, ES_MUTE_INVERSION_TYPE_PATH)` will result in the default mute set being selected rather than muted.
 * In most cases this is unintended.
* Consider calling `es_unmute_all_target_paths` before inverting target path muting. 
```

Reply:
But I still don't follow one thing; If the mute/unmute state doesn't start from a clean slate everytime a configure is called, and then we apply the file paths, aren't we now accumulating included paths?
I see that we clear `file_paths_` now, so that we don't include paths that were maybe present in the previous configuration file state, but we are not using any ES API to tell to exclude them again.

So if the configuration file contained a request to track `/tmp`, and later we change it to `/etc`, `file_paths_` will contain `/tmp`, but won't the ES client state contain the request to track both `/tmp` and `/etc` because at the second config refresh the only thing we asked it is to mute (include) `/etc`?

Reply:
Yes, that is correct. That exclusions for changes to `file_path` along with other improvements is coming in another PR.
For the sake of time, especially with trying to get this PR in this release to make it useable, I have opted to keep this PR small. For folks using this and ES functionality, there is also an implicit awareness, that they need to restart osquery for any ES config changes.

Reply:
I see, I didn't know that that was the intention beforehand, since the `configure` function gets called on config refresh.

Then I think my first suggestion is adding a problem, because there is still a chance of accumulating inclusion paths with configuration changes, and this can be unexpected for some, even if documented.
i think it's better that the "unexpected" is that changing `file_paths`, for now, doesn't react to any change,

Reply:
> there is still a chance of accumulating inclusion paths with configuration changes, and this can be unexpected for some, even if documented.

Yep, though I haven't been able to trigger it, but I agree. 

> i think it's better that the "unexpected" is that changing file_paths, for now, doesn't cause any change in monitored paths while osquery is running,

I have reverted to the original.